### ISSUE TITLE
[css-color-4][css-color-adjust-1] Shielding system colors to avoid fingerprinting?

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#5710]
The main privacy-related leak from css-color-4/css-color-adjust-1 is the way that the used value of the system colors are exposed. One way to mitigate this would be to define that the system colors always return a fixed set of values via CSSOM APIs, and that forcing colors does not affect getComptuedStyle(). That way the real used color is not Web-exposed. Do we want to do that?

The upside is that we shield the system colors from the Web: all browsers return the same values via getComputedStyle() all the time. The downside is that if the author was trying to do any interesting calculations to choose additional colors based on the user's preferred color palette, we're giving them bogus answers. (Also it's probably a bit more complicated to implement.)

**Comment by heycam** [comment]
> One way to mitigate this would be to define that the system colors always return a fixed set of values via CSSOM APIs, and that forcing colors does not affect getComptuedStyle(). That way the real used color is not Web-exposed. Do we want to do that?

We'd need to use these fixed values in canvas, too, at least.

What's the current compat status of the system colors? Would it be safe to use a fixed set of colors unconditionally, or perhaps dependent only on dark mode / high contrast mode, which can be detected with the MQs? (I seem to remember that at one point the system colors were deprecated, and then they were undeprecated.)

**Comment by svgeesus** [comment]
> (I seem to remember that at one point the system colors were deprecated, and then they were undeprecated.)

Right. They were all deprecated in CSS Color 3, then some of them un-deprecated in CSS Color 4

**Comment by svgeesus** [comment]
```
<dael> Rossen_: As a group are we comfortable with option 1? Objections to say no change on this and we'll document the fingerprinting exposure?
<dael> smfr: I need to talk to privacy people before I agree
<dael> Rossen_: In that case we'll leave it open as is. In both cases we need more data before we resolve
<dael> Rossen_: smfr can you follow up and we bring it back next week?
<dael> smfr: Sure.
<dael> Rossen_: Let's do that. We'll give the Apple folks a week and depending on how things go we may pick option 1
```

I know there was the seasonal festivities break but it has been more than a week so yes, lets get this resolved. The fingerprinting exposure is already mentioned in [Security and Privacy Considerations](https://drafts.csswg.org/css-color-4/#priv-sec). I'm therefore fine with option 1.

**Comment by sysrqb** [comment]
I'm providing a privacy review of this draft.

The information leakage in the forced-color property is quite concerning. As the feature is specified, the privacy and security risk is not only fingerprinting, but it is potentially revealing information about a user's health/physical condition. I understand the significant benefits in providing a mechanism for adjusting sites to meet the needs of the user, however it seems the current design allows a web page to abuse this information, too. I recommend considering options (2) or (3). Alternatively, the "correctness" of the response from getComputedStyle seems like functionality that could be gated behind a permission prompt.

Separately, regarding the color-scheme, I recommend limiting the addition of new schemes in the future.

**Comment by npdoty** [comment]
+1 that this is an information disclosure risk, not just a fingerprinting risk, especially if (as the IRC log above suggests), there are known direct cases of sites using getComputedStyle to detect whether the user has an accessibility mode turned on.

I would think that for color (as opposed to other properties that affect layout), lying in getComputedStyle is an effective mitigation, as in :visited/history sniffing.

In this case it might also require limitations on new media queries.

**Comment by fantasai** [comment]
This is a summary of the state of the CSSWG discussion:

The CSS system colors, in forced colors mode, are required to match the user-required colors so that the page can style itself in that color scheme. They also are required to change based on color-scheme, so that the default text and background colors match the user preference.

The used values of these colors are exposed via getComputedStyle(), which creates a privacy leak.

Note that the following information is already provided explicitly via Media Queries:
- Whether forced colors mode is on
- What color scheme the reader prefers
    
Thus the additional privacy leak presented by getComputedStyle() is not of a new type of information, but rather merely a greater amount of information for users that are already segmentable via MQs.

There are three ways forward here:
1. Continue to return the used valuse in getComputedStyle().
2. Return the system color keywords rather than the colors themselves.
3. Lie about the colors in getComputedStyle().

For the case of :visited, we took approach (3).
- First, because it was a much more severe data leak (of the user's browsing history).
- Secondly, because the impact of lying about :visited colors by returning :link colors is not very severe. Because :link and :visited are almost always styled very similarly in terms of color contrast, even if the author calculates some colors to use based on the returned :link color rather than the true :visited color, the text is most probably still readable.

But for the case of forced colors and color schemes, this approach is much more dangerous to the readability of the page. If the page believes itself to be rendered as dark text on a white background, but the system forces it to white text on a black background without, and the author calculcates custom color variants based on its own colors rather than the true colors, the page could become unreadable. Note that while the author can guess that this has happened for the page overall by combining the prefers-color-scheme media query with the forced-colors media query, these queries will only tell the author about the contrast nature of the Canvas and CanvasText colors, and not of button colors or other elements on the page, whose contrasts could be inverted.

Meanwhile the concern about option (2) is that, because of the existence of color schemes, all of the elements on the page are now expressed using system colors in the default UA stylesheet, whereas before they were all expressed as absolute colors. Because of this history of absolute colors returned by getComputedStyle(), the Web expects getComputedStyle() to return absolute colors for any element styled in its default colors: changing it to return keywords is therefore likely a change that we cannot make without breaking many pages.

The current conclusion of the CSSWG is thus that we believe that the risk of breaking pages (either via JS errors or by unreadable colors) for users employing forced colors is greater than the harm from increased fingerprinting entropy for those users.

We submit this analysis for review by Accessibility, Privacy, and the TAG.

**Comment by annevk** [comment]
Somewhat aside: I thought `:visited` wasn't solved yet? See #3012.

More on topic: [doesn't this leak through canvas as well](https://software.hixie.ch/utilities/js/canvas/?c.clearRect(0%2C%200%2C%20640%2C%20480)%3B%0Ac.fillStyle%20%3D%20%22Background%22%0Ac.fillRect(0%2C%200%2C%20640%2C%20640)%3B%0Aconsole.log(...c.getImageData(0%2C%200%2C%201%2C%201).data)%3B)? 2/3 won't work there.

I think part of the question here is how the OS allows customization of these. If these are color pickers of the `<input type=color>` variety, it seems highly likely users that use this kind of thing would make themselves fingerprintable.

**Comment by fantasai** [comment]
> More on topic: doesn't this leak through canvas as well? 2/3 won't work there.

@annevk Yes, as smfr [noted on the call](https://lists.w3.org/Archives/Public/www-style/2021Jan/0014.html) implementations would have to taint the canvas, which also means fingerprinting mitigations here would be very complex to implement as well.

> If these are color pickers of the <input type=color> variety

IIRC you _can_ choose your own colors (and for some users this may be necessary), but there are several pre-configured themes.

**Comment by svgeesus** [comment]
Given the [TAG response on this issue](https://github.com/w3ctag/design-reviews/issues/637#issuecomment-853079678) what is the next step? Do we need to evaluate the scale of Web-incompatible breakage if we return keywords? Or do a trial of the breakage from the "CSSOM lies about it" option?

**Comment by fantasai** [comment]
@svgeesus It looks like the TAG response lists the possible behaviors in the same preferential order as we did... but it doesn't seem like they're issuing any particular recommendation here and definitely their minutes show no discussion about the downsides we identified with each and no comparison of the options to determine the best realistic way forward. So afaict, they have no actionable feedback for us, and we should continue with the “status quo” option, given the listed (assumed probably fatal) downsides of the other options.

**Comment by svgeesus** [comment]
I agree, just wanted to be sure.

**Comment by svgeesus** [comment]
So that rationale should probably be added to the security and privacy appendix.

**Comment by tabatkins** [comment]
@svgeesus Added some language summarizing the reasoning from Elika's earlier comment, and linked back to here. That work for you?

**Comment by sysrqb** [comment]
> This is a summary of the state of the CSSWG discussion:
> 
> The CSS system colors, in forced colors mode, are required to match the user-required colors so that the page can style itself in that color scheme. They also are required to change based on color-scheme, so that the default text and background colors match the user preference.
> 
> The used values of these colors are exposed via getComputedStyle(), which creates a privacy leak.
> 
> Note that the following information is already provided explicitly via Media Queries:
> 
>     * Whether forced colors mode is on
> 
>     * What color scheme the reader prefers
> 
> 
> Thus the additional privacy leak presented by getComputedStyle() is not of a new type of information, but rather merely a greater amount of information for users that are already segmentable via MQs.
> 

Yes, I agree, but we're looking for solutions that limit harm (not necessarily eliminate it). However, I do agree with you now that lying is not a good solution and my original concern was misplaced. Thank you for your detailed response, it was very helpful.

Specifically, just so my thoughts on this are clear: as mentioned above, in the three proposed options the state of `forced-colors` would not be modified, therefore the privacy leak is inherent in this accessibility feature and it has nothing to do with the resulting styling (via gCS). The modified styling when `forced-colors` is enabled is likely UA/OS specific, but that information is already available to JS from other sources.

In the future I would like to see a privacy-preserving MQ mechanism that doesn't leak so much information about the UA/OS, but I know that this draft should not be a critique on MQ in general, so we can move on from this.

> The current conclusion of the CSSWG is thus that we believe that the risk of breaking pages (either via JS errors or by unreadable colors) for users employing forced colors is greater than the harm from increased fingerprinting entropy for those users.

I agree this is a reasonable assessment.

**Comment by svgeesus** [comment]
> Added some language summarizing the reasoning from Elika's earlier comment, and linked back to here. That work for you?

Perfect
